### ISSUE TITLE
Cannot specify index name using path setting on Spark SQL 3.0 temp tables

### ISSUE THREAD

**Issue by jbaiera** [elastic/elasticsearch-hadoop#1593]
The Spark SQL integration support allows specifying the Elasticsearch index resource name [a number of ways](https://www.elastic.co/guide/en/elasticsearch/hadoop/current/spark.html#spark-data-sources). One of those configurations is by using the `path` setting when creating a Dataset. This path setting can be specified in raw SQL as well. 

We encountered a problem with Spark during the 3.0 upgrade: When using the `CREATE TEMP TABLE`  statement in SQL form, the path setting was being accepted but not passed to the underlying integration code. Based on the logic in `org.apache.spark.sql.catalyst.parser.AstBuilder#cleanTableOptions`, the `path` setting is now explicitly filtered out of the table options. This path is set to be the _location_ of the table, which is also specifiable through the `LOCATION` directive in the create table syntax.

The original logic for this parsing in 2.x is back in `org.apache.spark.sql.execution.SparkSqlParser#visitCreateTable`. Like in 2.x, it was possible to specify `LOCATION` or `path`, but still impossible to use both. 

It seems that in 3.0 the `path` option is still passed into the Datasource when creating a non-temporary table. The table location is rendered into an option entry and passed into the data source. See `org.apache.spark.sql.execution.command.CreateDataSourceTableCommand#run`

I feel like this inconsistency may be a bug with Spark. I'll see if I can raise the issue with the Spark Community.

**Comment by masseyke** [comment]
@jbaiera I'm wondering if spark has fixed this because I was unable to reproduce the problem. Here's what I tried in spark 3.2.1 using es 7.17.0 and elasticsearch-spark 30_2.12-7.17.0 on a really simple index called `test_index` that I'd put a single document into:
```
val sqlContext = new org.apache.spark.sql.SQLContext(sc)
val df = sqlContext.read.format("es").load("test_index")
df.show
val x = sqlContext.sql("CREATE TEMPORARY TABLE trips1 USING org.elasticsearch.spark.sql OPTIONS (path 'test_index')")
val y = sqlContext.sql("CREATE TEMPORARY TABLE trips2 USING org.elasticsearch.spark.sql LOCATION 'test_index'")
val z = sqlContext.sql("CREATE TEMPORARY TABLE trips3 USING org.elasticsearch.spark.sql OPTIONS(resource 'test_index')")
sqlContext.sql("select * from trips1").show
sqlContext.sql("select * from trips2").show
sqlContext.sql("select * from trips3").show
```
All of that worked as expected. I also pulled down spark 2.4.3 and elasticsearch-spark 20_2.11-7.17.0 and tried all of the above. Oddly, `trips2` broke in that case (the one that uses LOCATION) with this stack trace:
```
org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: resource must be specified for Elasticsearch resources.
  at org.elasticsearch.spark.sql.DefaultSource$$anonfun$params$1$$anonfun$apply$1.apply(DefaultSource.scala:188)
  at org.elasticsearch.spark.sql.DefaultSource$$anonfun$params$1$$anonfun$apply$1.apply(DefaultSource.scala:188)
  at scala.collection.MapLike$class.getOrElse(MapLike.scala:128)
  at scala.collection.AbstractMap.getOrElse(Map.scala:59)
  at org.elasticsearch.spark.sql.DefaultSource$$anonfun$params$1.apply(DefaultSource.scala:188)
  at org.elasticsearch.spark.sql.DefaultSource$$anonfun$params$1.apply(DefaultSource.scala:188)
  at scala.collection.MapLike$class.getOrElse(MapLike.scala:128)
  at scala.collection.AbstractMap.getOrElse(Map.scala:59)
  at org.elasticsearch.spark.sql.DefaultSource.params(DefaultSource.scala:187)
  at org.elasticsearch.spark.sql.DefaultSource.createRelation(DefaultSource.scala:97)
  at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:318)
  at org.apache.spark.sql.execution.datasources.CreateTempViewUsing.run(ddl.scala:93)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
  at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
  at org.apache.spark.sql.Dataset$$anonfun$6.apply(Dataset.scala:194)
  at org.apache.spark.sql.Dataset$$anonfun$6.apply(Dataset.scala:194)
  at org.apache.spark.sql.Dataset$$anonfun$53.apply(Dataset.scala:3364)
  at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
  at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3363)
  at org.apache.spark.sql.Dataset.<init>(Dataset.scala:194)
  at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:79)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:642)
  at org.apache.spark.sql.SQLContext.sql(SQLContext.scala:694)
  ... 49 elided
```

**Comment by masseyke** [comment]
The error I ran into might be related to https://github.com/apache/spark/pull/30014. According to that ticket it was introduced in spark 2.4. 
I just added an integration test locally for spark 1.x, 2.x, and 3.x, and confirmed that using `LOCATION` works in spark 1.6 and 3.x but fails in 2.x. I'm not sure there's anything we can do about it for 2 if it's a spark bug that has since been fixed. But I'm also not sure that this is the problem you were originally describing.

**Comment by masseyke** [comment]
The ticket I reference above was added in spark 3.0.2. I just confirmed that we get the stack trace I posted above in spark 3.0.1 but it succeeds in spark 3.0.2, so I'm pretty sure that's it. And I don't think we can do much about it because the parameters map we're given in our DataSource.createRelation just has no parameters at all. So the workarounds are to either just don't use that syntax, or upgrade to spark 3.0.2 or newer.
